source/opt/cfg_cleanup_pass.h

```diff
@@ -0,0 +1,88 @@
+// Copyright (c) 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef LIBSPIRV_OPT_CFG_CLEANUP_PASS_H_
+#define LIBSPIRV_OPT_CFG_CLEANUP_PASS_H_
+
+#include "function.h"
+#include "module.h"
+#include "mem_pass.h"
+
+namespace spvtools {
+namespace opt {
+
+class CFGCleanupPass : public MemPass {
+ public:
+  CFGCleanupPass() = default;
+  const char* name() const override { return "cfg-cleanup"; }
+  Status Process(ir::Module*) override;
+
+ private:
+  // Call all the cleanup helper functions on |func|.
+  bool CFGCleanup(ir::Function* func);
+
+  // Remove all the unreachable basic blocks in |func|.
+  bool RemoveUnreachableBlocks(ir::Function* func);
+
+  // Remove the block pointed by the iterator |*bi|. This also removes
+  // all the instructions in the pointed-to block.
+  void RemoveBlock(ir::Function::iterator *bi);
+
+  // Initialize the pass.
+  void Initialize(ir::Module* module);
+
+  // Remove the result from |inst| from every Phi instruction reached by
+  // |inst|. The instruction is coming from basic block |block|.
+  void RemoveFromReachedPhiOperands(const ir::BasicBlock& block,
+                                    ir::Instruction* inst);
+
+  // Remove Phi operand (|rm_arg_id|, |rm_block_id|) from |phi|.
```

Comment:
This needs slightly more explanation?
```
Remove Phi operand involving either |rm_arg_id| or |rm_block_id|:
- Remove any pair (var, |rm_block_id|)
- Otherwise, replace (|rm_arg_id|, block) with ( U, block ) where U is the undefined value with same type as |rm_arg_id|
```

Reply:
It would also be handy if you allowed |rm_arg_id| to be zero.  Then you can use it for the cleanup of OpLabel mentioned in phis, e.g. where the value is some OpConstant.